File: src/main/java/org/inksnow/ankh/core/hologram/hds/HdsHologramContent.java
package org.inksnow.ankh.core.hologram.hds;

import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import me.filoghost.holographicdisplays.api.hologram.HologramLines;
import me.filoghost.holographicdisplays.api.hologram.line.ItemHologramLine;
import me.filoghost.holographicdisplays.api.hologram.line.TextHologramLine;
import org.bukkit.inventory.ItemStack;
import org.inksnow.ankh.core.api.hologram.HologramContent;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HdsHologramContent implements HologramContent {

    public interface LineEntry {}

    @AllArgsConstructor
    @NoArgsConstructor
    private static class Text implements LineEntry {
        public String content;
    }

    @AllArgsConstructor
    @NoArgsConstructor
    private static class Item implements LineEntry {
        public ItemStack item;
    }

    private final List<LineEntry> lines;

    public HdsHologramContent(List<LineEntry> lines) {
        this.lines = Collections.unmodifiableList(lines);
    }

    public void applyToLines(HologramLines hdsLines) {
        while (hdsLines.size() > lines.size()) {
            hdsLines.remove(hdsLines.size() - 1);
        }

        for (int i = 0; i < hdsLines.size(); i++) {
            if (hdsLines.get(i) instanceof Text) {
                if (hdsLines.size() <= i) {
                    hdsLines.appendText(((Text) hdsLines.get(i)).content);
                } else if (hdsLines.get(i) instanceof TextHologramLine) {
                    ((TextHologramLine) hdsLines.get(i)).setText(((Text) hdsLines.get(i)).content);
                } else {
                    hdsLines.insertText(i, ((Text) hdsLines.get(i)).content);
                    hdsLines.remove(i);
                }
            } else if (hdsLines.get(i) instanceof Item) {
                if (hdsLines.size() <= 1) {
                    hdsLines.appendItem(((Item) hdsLines.get(i)).item);
                } else if (hdsLines.get(i) instanceof ItemHologramLine) {
                    ((Item) hdsLines.get(i)).item = ((Item) hdsLines.get(i)).item;
                } else {
                    hdsLines.insertItem(i, ((Item) hdsLines.get(i)).item);
                    hdsLines.remove(i);
                }
            }
        }
    }

    public static class Builder implements HologramContent.Builder {

        private final ArrayList<LineEntry> lines = new ArrayList<>();

        @Override
        public HologramContent.Builder appendContent(String content) {
            lines.add(new Text(content));
            return this;
        }

        @Override
        public HologramContent.Builder appendItem(ItemStack item) {
            lines.add(new Item(item));
            return this;
        }

        @NotNull
        @Override
        public HologramContent.Builder getThis() {
            return this;
        }

        @NotNull
        @Override
        public HologramContent build() {
            return new HdsHologramContent(lines);
        }
    }
}

File: src/main/kotlin/bot/inker/ankh/core/hologram/hds/HdsHologramContent.kt
package bot.inker.ankh.core.hologram.hds

import bot.inker.ankh.core.api.hologram.HologramContent
import me.filoghost.holographicdisplays.api.hologram.HologramLines
import me.filoghost.holographicdisplays.api.hologram.line.ItemHologramLine
import me.filoghost.holographicdisplays.api.hologram.line.TextHologramLine
import org.bukkit.inventory.ItemStack
import java.util.*

class HdsHologramContent(
  lines: List<LineEntry>,
) : HologramContent {
  private val lines = Collections.unmodifiableList(ArrayList(lines))

  fun applyToLines(hdsLines: HologramLines) {
    // remove more entries
    while (hdsLines.size() > lines.size) {
      hdsLines.remove(hdsLines.size() - 1)
    }
    /*for (i in 0 until lines.size) {
      val lineEntry = lines[i]
      if (lineEntry is Text) {
        if(hdsLines.size() <= i) { // append new entry
          hdsLines.appendText(lineEntry.content)
        } else if(hdsLines[i] is TextHologramLine){ // hds line with same type, just update
          (hdsLines[i] as TextHologramLine).text = lineEntry.content
        } else { // hds line with different type, insert new one and remove it
          hdsLines.insertText(i, lineEntry.content)
          hdsLines.remove(i)
        }
      }else if(lineEntry is Item) {
        if(hdsLines.size() <= i){ // append new entry
          hdsLines.appendItem(lineEntry.item)
        }else if (hdsLines[i] is ItemHologramLine) { // hds line with same type, just update
          (hdsLines[i] as ItemHologramLine).itemStack = lineEntry.item
        }else{ // hds line with different type, insert new one and remove it
          hdsLines.insertItem(i, lineEntry.item)
          hdsLines.remove(i)
        }
      }
    }*/
    // refactor the code above with sequence api for readability
    // aims to avoid if and else
    lines.forEachIndexed { index, lineEntry ->
      if (hdsLines.size() <= index) { // append new entry
        when (lineEntry) {
          is Text -> hdsLines.appendText(lineEntry.content)
          is Item -> hdsLines.appendItem(lineEntry.item)
        }
      } else { // hdsLines.size() > index
        val hdsLine = hdsLines[index]
        when {
          lineEntry is Text && hdsLine is TextHologramLine -> hdsLine.text = lineEntry.content
          lineEntry is Item && hdsLine is ItemHologramLine -> hdsLine.itemStack = lineEntry.item
          else -> {
            when (lineEntry) {
              is Text -> hdsLines.insertText(index, lineEntry.content)
              is Item -> hdsLines.insertItem(index, lineEntry.item)
            }
            hdsLines.remove(index)
          }
        }
      }
    }
  }

  class Builder : HologramContent.Builder {
    override fun getThis() = this
    private val lines = ArrayList<LineEntry>()

    override fun build() = HdsHologramContent(lines)

    override fun appendContent(content: String) = apply {
      lines.add(Text(content))
    }

    override fun appendItem(item: ItemStack) = apply {
      lines.add(Item(item))
    }
  }

  sealed interface LineEntry
  class Text(val content: String) : LineEntry
  class Item(val item: ItemStack) : LineEntry
}